src/units/abilities.cpp

```diff
@@ -1171,14 +1171,216 @@ unit_ability_list attack_type::get_special_ability(const std::string& ability) c
+{
+	return (get_adj_ability_bool(special, tag_name, dir, loc, from) && ability_affects_weapon(special, weapon, false) && ability_affects_weapon(special, opp_weapon, true));
+}
+
+bool attack_type::check_self_abilities(const config& cfg, const std::string& special, std::set<std::string> included_tags) const
+{
+	return check_self_abilities_impl(shared_from_this(), other_attack_, cfg, self_, self_loc_, AFFECT_SELF, special, included_tags, true);
+}
+
+bool attack_type::check_self_abilities_impl(const_attack_ptr self_attack, const_attack_ptr other_attack, const config& special, unit_const_ptr u, const map_location& loc, AFFECTS whom, const std::string& tag_name, std::set<std::string> included_tags, bool leader_bool)
+{
+	if(tag_name == "leadership" && leader_bool){
+		if((*u).get_self_ability_bool_weapon(special, tag_name, loc, self_attack, other_attack)) {
+			return true;
+		}
+	}
+	if(included_tags.count(tag_name) != 0){
+		if((*u).get_self_ability_bool(special, tag_name, loc) && special_active_impl(self_attack, other_attack, special, whom, tag_name, true, "filter_student")) {
+			return true;
+		}
+	}
+	return false;
+}
+
+bool attack_type::check_adj_abilities(const config& cfg, const std::string& special, int dir, const unit& from, std::set<std::string> included_tags) const
+{
+	return check_adj_abilities_impl(shared_from_this(), other_attack_, cfg, self_, from, dir, self_loc_, AFFECT_SELF, special, included_tags, true);
+}
+
+bool attack_type::check_adj_abilities_impl(const_attack_ptr self_attack, const_attack_ptr other_attack, const config& special, unit_const_ptr u, const unit& from, int dir, const map_location& loc, AFFECTS whom, const std::string& tag_name, std::set<std::string> included_tags, bool leader_bool)
 {
-	bool abil_bool = get_special_bool(ability);
-	unit_ability_list abil = list_ability(ability);
-	if(!abil.empty()) {
-		abil_bool = true;
+	if(tag_name == "leadership" && leader_bool){
+		if((*u).get_adj_ability_bool_weapon(special, tag_name, dir, loc, from, self_attack, other_attack)) {
+			return true;
+		}
 	}
-	return abil_bool;
+	if(included_tags.count(tag_name) != 0){
+		if((*u).get_adj_ability_bool(special, tag_name, dir, loc, from) && special_active_impl(self_attack, other_attack, special, whom, tag_name, true, "filter_student")) {
+			return true;
+		}
+	}
+	return false;
+}
+/**
+ * Returns whether or not @a *this has a special ability with a tag or id equal to
+ * @a special. the Check is for a special ability
+ * active in the current context (see set_specials_context), including
+ * specials obtained from the opponent's attack.
+ */
+bool attack_type::get_special_ability_bool(const std::string& special, bool special_id, bool special_tags) const
+{
+	static std::set<std::string> included_tags{"damage", "chance_to_hit", "berserk", "swarm", "drains", "heal_on_hit", "plague", "slow", "petrifies", "firststrike", "poison"};
```

Comment:
Making it static doesn't change the fact that this set of values is defined in multiple places.  `{"damage", "chance_to_hit", "berserk", "swarm", "drains", "heal_on_hit", "plague", "slow", "petrifies", "firststrike", "poison"}` should appear *once* in all the code rather than in three separate places like it does now.

One other option would be to use MAKE_ENUM to define the possible values and then check for matches with the `string_to_enum(str_val, enum_default)` method.

Reply:
MAKE_ENUM? imust define a case for each weapon_type?

Reply:
You would make a NOT_FOUND enum value or similar, then do `if(string_to_enum(str_val, NOT_FOUND) == NOT_FOUND)`

Reply:
i declared std::set<std::string> included_tags in unit.hpp instead.